feat(CodeSnippet): set closed and expanded max number of rows

guigueb opened the pull request:
defaulted the height to show everything (100%)
added two new props maxClosedNumberOfRows and maxExpandedNumberOfRows
added logic using the new props to determine when to show moreLessBtn
added maxHeight style based on new props

Closes #

https://github.com/carbon-design-system/carbon/issues/7579

#### Testing / Reviewing

See [7579](https://github.com/carbon-design-system/carbon/issues/7579) for the tests to verify this PR.

Tested:

MacOS - Chrome, FireFox, and Safari
Win10 - Chrome, FireFox, and Edge

The margin fix described in [7582](https://github.com/carbon-design-system/carbon/pull/7582) was not done based on comments in [7826](https://github.com/carbon-design-system/carbon/pull/7826).

Any issues with (scrollbar, WCAG 1.4.10 compliance, minHeight, and overFlow indicators) have been discussed in [7826](https://github.com/carbon-design-system/carbon/pull/7826) will be addressed in these future PRs.
[8056](https://github.com/carbon-design-system/carbon/issues/8056) CodeSnippet: there should be a way to set the min number of rows when in closed and expanded mode
[7574](https://github.com/carbon-design-system/carbon/issues/7574) CodeSnippet: missing horizontal scrollbar when type is multi and there is a long row of data
[8057](https://github.com/carbon-design-system/carbon/issues/8057) CodeSnippet: the scrollable content needs to be WCAG 1.4.10 compliant
[8059](https://github.com/carbon-design-system/carbon/issues/8059) CodeSnippet: Remove or improve the overflow indicators

Commit:
feat(codesnippet): set closed and expanded max number of rows

defaulted the height to show everything (100%)
added two new props maxClosedNumberOfRows and maxExpandedNumberOfRows
added logic using the new props to determine when to show moreLessBtn
added maxHeight style based on new props

guigueb commented:
Where can I see the conformance list for the platforms and browsers Carbon supports?
I likely missed it reading the Carbon Design System pages.

The reason being... I viewed CodeSnippet on my Samsung Tablet and the number of lines displayed does not match the max values given.
Although, if this is a problem its beyond this PR as it wasn't displaying 15 before this PR.

emyarod reviewed (commented):
Carbon targets evergreen browsers, but that would include the mobile variants so I think the issue with the incorrect number of lines being shown on mobile browsers should be addressed together

guigueb commented:
@emyarod
I just retested and I am not able to reproduce the issue.
I believe the pixel count per row was off, and when I adjusted rowHeightInPixels from 18 to 16 it addressed the issue.

You mention evergreen browsers... can you point me to the conformance text on the Carbon Design System pages that mentions this?

The list of evergreen browsers (I learnt something) is huge and my ability to test them is limited (I can do mac, win, Samsung Tab S2, and a very old chromebook) so hard to tell if it works everywhere.

emyarod commented:
@guigueb [here](https://github.com/carbon-design-system/carbon/tree/main/config/browserslist-config-carbon) is the Carbon browserslist

emyarod commented on packages/react/src/components/CodeSnippet/CodeSnippet.js:
```diff
@@ -147,6 +162,23 @@ function CodeSnippet({
+        };
+      }
+    } else {
+      if (maxClosedNumberOfRows > 0) {
```
```suggestion
    } else  if (maxClosedNumberOfRows > 0) {
```

this does not need to be nested right?

guigueb replied:
It does not, for this PR.
The code is a set up for the next PR for #8056 dealing with minHeight, it will need to be nested (or at least reworked).

emyarod reviewed (commented):
also requesting ux/visual review for this feature

tw15egan commented:
Since this PR introduces the ability to set the max number of rows when expanded to _less_ than the total number of lines, I think we'll also need to add a selector that allows vertical scrolling, something like

```scss
  .#{$prefix}--snippet--multi.#{$prefix}--snippet--expand
    .#{$prefix}--snippet-container {
    overflow-y: auto;
  }
```

guigueb commented:
I completely agree that overflow(scrollbars) should be taken into account for this.
That was attempted in #7826 but scrollbars are not directly related to adding the max props (although related/exacerbated, scrolling is an existing issue with large 100% height CodeSnippets - which is what I'm working on fixing) and introduces issues with WCAG 1.4.10 compliance.
As such they should be dealt with in #7574 and #8057.

tw15egan commented:
@guigueb in the current implementation, when the content is expanded, it shows every line of code. If this PR introduces the ability to show less than the max number, you need to be able to scroll to the bottom to view all the code, and should not be handled in a separate PR. 

Example: In the preview, set the max expanded to `16`. When you hit expand, it only expands 1 line, and the rest of the code cannot be scrolled to

guigueb commented:
I aware of the issue and agree, I tried to address this with my other PR - and was told to stick to the props.
I'll add changes for overflow back into this one, keep in mind it will be non-WCAG 1.4.10 compliant.

With current CodeSnippet showing 100% if there are 20000 lines you see only what you can... and the rest of the code cannot be scrolled to (same issue different number of max rows).
I suppose you can work around this by putting it in a scrollable container but that only causes other issues.
The vertical scroll issue isn't new but it is easer to see, with the limit.

tw15egan commented:
@guigueb it seems like we just need to add the line I suggested to fix that issue. The current implementation will push all other content down, no matter the number of lines. I'm not seeing any "unreachable" code

![scroll](https://user-images.githubusercontent.com/11928039/111698512-2657f280-87f4-11eb-9ff0-5f3fe56d4f66.gif)

Commit:
feat(CodeSnippet): apply auto scroll to snippet-mult

allow snippet-mult + snippet-container to auto scroll, both x and y
remove scroll/overflow from nested snippet-expand, pre, and code conditions
this will show the horizontal scroll when the snippet is closed or expanded
it is a behavour change/bug fix, it will address 7574

Commit:
feat(CodeSnippet): add background to snippet buttons

add background to buttons to stop scrollbars from stomping on them

Commit:
feat(CodeSnippet): revert to only adding overflow-y

changing to overflow - from -x/-y - on the snippet-container
caused issues with scrollbars colliding with the buttons.
for now only add overflow-y and address overflow clean up and
scrollbars/button collisions in #7574

tw15egan committed:
Merge branch 'main' into feat-(codesnippet)-set-closed-and-expanded-max-number-of-rows

tw15egan commented on packages/components/src/components/code-snippet/_code-snippet.scss:
```diff
@@ -202,16 +202,16 @@
     .#{$prefix}--snippet-container {
-    max-height: 100%;
     padding-bottom: $spacing-05;
+    overflow-y: auto;
```
Can we move this to L206?

guigueb replied:
Sure can - apply it for both closed and expanded modes

Commit:
feat(CodeSnippet): move overflow-y placement

apply it for both closed and expanded modes

tw15egan reviewed (approved):
Thanks for working on this! LGTM 👍 ✅

tay1orjones commented on packages/react/src/components/CodeSnippet/CodeSnippet.js:
```diff
@@ -253,6 +286,16 @@ CodeSnippet.propTypes = {
+  /**
+   * Specify the maximum number of rows to be shown when in closed view
+   */
+  maxClosedNumberOfRows: PropTypes.number,
```
Would it make sense to use to "collapsed" instead of "closed'? I think there's precedent elsewhere in the component for the concept of it "expand"ing which would point to using "collapse"

```suggestion
  maxCollapsedNumberOfRows: PropTypes.number,
```

guigueb replied:
The name can be changed if you want.
I went with closed, because comments in the scss used closed.
Let me know.

guigueb replied:
changed

tay1orjones reviewed (commented):
Just one small piece of feedback on prop naming. Looks great otherwise!

Commit:
feat(CodeSnippet): changed closed to collapsed

changed closed to collapsed

Commit:
feat(CodeSnippet): update snapshot

update snapshot

tw15egan commented on packages/react/src/components/CodeSnippet/CodeSnippet-story.js:
```diff
@@ -43,6 +49,14 @@ const props = () => ({
   ariaLabel: text('ARIA label', 'Container label'),
   wrapText: boolean('Wrap text (wrapText)', true),
+  maxClosedNumberOfRows: number(
+    'maxClosedNumberOfRows: Specify the maximum number of rows to be shown when in closed view',
```
```suggestion
  maxCollapsedNumberOfRows: number(
    'maxCollapsedNumberOfRows: Specify the maximum number of rows to be shown when in closed view',
```

guigueb replied:
Crap - missed the story.
Done now.

Commit:
feat(CodeSnippet): rename closed to collapsed in the story

rename closed to collapsed in the story

Commit:
Merge remote-tracking branch 'upstream/main' into feat-(codesnippet)-set-closed-and-expanded-max-number-of-rows

guigueb committed:
Merge branch 'main' into feat-(codesnippet)-set-closed-and-expanded-max-number-of-rows

aagonzales reviewed (approved):
Looks like it's working and looking correct to me!

kodiakhq[bot] committed:
Merge branch 'main' into feat-(codesnippet)-set-closed-and-expanded-max-number-of-rows
